Preliminary concurrency support with mioco.

nhoad opened the pull request:
mioco runs on top of mio and uses coroutines, so they're quite comfortable to write.

However it does mean we have to run on nightly, which is a thing to consider.

Also, performance - I don't have any hard numbers, but the concurrency benchmarks given by the author are quite impressive: https://github.com/dpc/mioco/blob/master/BENCHMARKS.md

368k requests a second with httparse is pretty impressive. That's with 100 client connections as well, which would translate to 100 threads in the old version, which wasn't exactly scalable.

Before we can come up with any realistic numbers for octopus, we need to implement response parsing from upstream - at the moment it's just a dumb tcp tunnel.

Commit:
Preliminary concurrency support with mioco.

mioco runs on top of mio and uses coroutines for concurrency, so they're
quite comfortable to write.

However it does we run on nightly, which is a thing to consider.

Commit:
Remove some leftover debugging.

miselin commented on src/main.rs:
```diff
@@ -1,129 +1,41 @@
-fn main() {
-    let port = 8000;
+            mioco::spawn(move || -> io::Result<()> {
+                handle_client(conn);
```
Why not make handle_client return a Result here? Does it blow up the loop?

nhoad replied:
Good point! I'll work on getting that going.

miselin commented on src/server.rs:
```diff
@@ -0,0 +1,100 @@
+        },
+        _ if len == capacity => {
+            println!("buffer is at capacity, growing {} -> {}", len, len + resize_amount);
+            buffer.resize(len + resize_amount, 0);
```
While this is being changed, maybe we should consider adding a reserve() call for len \* 2 before or after the call to resize(), under the assumption that we'll most likely read more . If the function is likely to be going away, though, we can probably leave this be.

nhoad replied:
I started writing some tests for this and I found all sorts of awful edge cases, so I've greatly simplified it - performance may suffer a bit, but at least it's bugfree now!

miselin commented:
Looks good. A couple notes but no blockers. Let's work towards getting some benchmarks up and running - I have a hunch that it'll be easier with mioco.

Commit:
Hugely simplify what read_into_buffer does.

Managing the write position state between calls is extremely difficult,
so we'll go with the potentially less performant version for the time
being and just read into a separate buffer instead.

Also add some tests.

Commit:
Make upstream connection also use mioco.

Also introduce some error handling for upstream connections, and
retries. Sadly this stops it working with domains for the time being,
we'll need a DNS solution soon.

Commit:
Pointless reassign removed.

Commit:
Humongous code reshuffle.

This does the following:

    * moves octopus::{request,headers,server} into octopus::http
      modules.

    * Moves upstream connection logic into octopus::http::client.

    * Moves Headers and Request serialization to using the
      std::convert::Into trait.

    * Introduces an as-yet-unused http::reply::Reply object.

    * Moves listening server code into a Server struct.

Commit:
Make handle_client return a Result.

Commit:
Stop switching between 65535 and 65536

Commit:
Fix the lifetime insanity of Headers, Request and Reply objects.

Replacing the &str types for Header.name, Request.method and
Reply.reason with String is worth not having to deal with lifetimes for
the time being.

This also allows parsing to be broken out into a separate function, and
remove the need for the special scoping in handle_request.

miselin commented:
Looks ready to submit to me. I'll look into building some benchmarks once this goes in.

nathan-hoad commented:
Awesome, cheers! I've got another change and some tests I want to get pushed a little bit later today, will poke you after that for another review.

Commit:
Handle request parsing errors.

As a result, throw away our url_is_relative function, instead letting
rust-url do its job.

nathan-hoad commented:
@miselin If the lifetime changes and parsing error handling looks good to you, I'd like to merge this in - I've got some other dependent stuff for doing DNS resolution and response parsing that I'd like to get started on :)

miselin commented:
Looks good to merge so we can iterate on it further.
